src/sumologic-context-manager/events.ts

```diff
@@ -0,0 +1,192 @@
+  const wrappedOnListeners = new WeakMap<OnListener, OnListener>();
+
+  const patchOnProperty = <O extends Object>(
+    object: O,
+    property: string,
+  ): void => {
+    const descriptor = Object.getOwnPropertyDescriptor(object, property);
+    if (!descriptor) {
+      const proto = Object.getPrototypeOf(object);
+      if (!proto) return;
+      return patchOnProperty(proto, property);
+    }
+    wrapWithToString(
+      descriptor,
+      'get',
+      (original) =>
+        function (this: unknown) {
+          const listener = original!.call(this);
+          return wrappedOnListeners.get(listener) || listener;
+        },
+    );
+    wrapWithToString(
+      descriptor,
+      'set',
+      (original) =>
+        function (this: unknown, listener: OnListener | null) {
+          let wrappedListener = listener;
+          if (listener) {
+            wrappedListener = contextManager.bind(
+              contextManager.active(),
+              listener,
+            );
+            wrappedOnListeners.set(wrappedListener, listener);
+          }
+          original!.call(this, wrappedListener);
+        },
+    );
+    Object.defineProperty(object, property, descriptor);
+  };
+
+  const patchOnProperties = <O extends Object>(
+    object: O,
+    properties: string[],
+  ): void => {
+    properties.forEach((property) => {
+      patchOnProperty(object, `on${property}`);
+    });
+  };
+
+  const wrappedEventListeners = new WeakMap<
+    EventListenerOrEventListenerObject,
+    EventListenerOrEventListenerObject
+  >();
+
+  wrapWithToString(
+    EventTarget.prototype,
+    'addEventListener',
+    (original) =>
+      function (this: EventTarget, ...args) {
+        const listener = args[1];
+        if (listener) {
+          let wrappedListener = wrappedEventListeners.get(listener);
+          if (!wrappedListener) {
+            if (typeof listener === 'function') {
+              wrappedListener = contextManager.bind(
+                contextManager.active(),
+                listener,
+              );
+            } else if (listener && typeof listener.handleEvent === 'function') {
+              wrappedListener = {
+                ...listener,
+                handleEvent: contextManager.bind(
+                  contextManager.active(),
+                  listener.handleEvent,
+                ),
+              };
+            }
+          }
+          if (wrappedListener) {
+            wrappedEventListeners.set(listener, wrappedListener);
+            args[1] = wrappedListener;
+          }
+        }
+        return original.apply(this, args);
+      } as typeof EventTarget.prototype.addEventListener,
+  );
+
+  wrapWithToString(
+    EventTarget.prototype,
+    'removeEventListener',
+    (original) =>
+      function (this: EventTarget, ...args) {
+        const listener = args[1];
+        if (listener) {
+          const wrappedListener = wrappedEventListeners.get(listener);
```

Comment:
What do you mean? `wrappedEventListeners` is a `WeakMap`.

Reply:
Yes, but there may be cases where `listener` is still available but it's not attached as a listener anymore; it will stay in the WeakMap potentially indefinitely then.

Maybe that's an edge case that we don't need to care. Plus, there are advantages as in React hooks there's a pattern to keep calling `addEventListener` & `removeEventListener` on each re-render in a `useEffect` hook and that way we don't need to re-wrap.

Reply:
Ok. Now I get it. I think it's kind of edge-case and it's not a leak, just an increased memory consumption, so let's leave it as it is, because it's much easier to maintain.